data_juicer/ops/deduplicator/ray_bts_minhash_deduplicator.py

```diff
@@ -0,0 +1,575 @@
+                tokens = self.tokenizer.encode(text, out_type=str)
+                return {
+                    str.encode(''.join(tokens[i:i + self.window_size]))
+                    for i in range(len(tokens) - self.window_size)
+                }
+        else:
+            raise NotImplementedError(
+                f'Unimplemented tokenization method [{self.tokenization}]')
+        self.tokenization_func = tokenization_func
+
+        # about deduplication
+        self.num_permutation = num_permutations
+        self.jaccard_threshold = jaccard_threshold
+        self.num_bands = num_bands
+        self.num_rows_per_band = num_rows_per_band
+
+        # initialize deduplication parameters
+        # check number of bands and rows
+        if self.num_bands is None or self.num_rows_per_band is None:
+            self.num_bands, self.num_rows_per_band = optimal_param(
+                self.jaccard_threshold,
+                self.num_permutation,
+            )
+
+        # compute hash ranges and create hash tables
+        self.hash_ranges = [(i * self.num_rows_per_band,
+                             (i + 1) * self.num_rows_per_band)
+                            for i in range(self.num_bands)]
+
+        # generate permutations
+        gen = np.random.RandomState(seed=42)
+        self.perm_a, self.perm_b = np.array(
+            [(
+                gen.randint(1, MERSENNE_PRIME, dtype=np.uint64),
+                gen.randint(0, MERSENNE_PRIME, dtype=np.uint64),
+            ) for _ in range(self.num_permutation)],
+            dtype=np.uint64,
+        ).T
+
+        if union_find_parallel_num == 'auto':
+            union_find_parallel_num = int(ray.cluster_resources().get('CPU') /
+                                          2)
+        else:
+            union_find_parallel_num = int(union_find_parallel_num)
+
+        self.max_pending_edge_buffer_task = max_pending_edge_buffer_task
+        self.num_edge_buffer_task_returns = num_edge_buffer_task_returns
+        self.max_pending_filter_tasks = max_pending_filter_tasks
+        self.num_filter_task_returns = num_filter_task_returns
+        self.merge_batch_size = min(merge_batch_size, union_find_parallel_num)
+
+        logger.info(f'union_find_parallel_num = {union_find_parallel_num}')
+        self.union_find_parallel_num = union_find_parallel_num
+        self.union_threshold = union_threshold
+        self.remote_edge_buffers = [
+            EdgeBuffer.remote() for _ in range(self.union_find_parallel_num)
+        ]
+        self.union_find_list = [
+            BTSUnionFind.remote(
+                self.union_threshold,
+                self.union_find_parallel_num,
+                i,
+                self.remote_edge_buffers,
+                self.max_pending_edge_buffer_task,
+                self.num_edge_buffer_task_returns,
+            ) for i in range(self.union_find_parallel_num)
+        ]
+
+        self.tmp_file_name = os.path.join(os.getcwd(), tmp_file_name,
```

Comment:
Moving the temporary file to a sub-directory of the output directory maybe better.

Reply:
fixed